File: objc/Project8/Project8/PlayScene.m
//
//  PlayScene.m
//  Project8
//
//  Created by Jinwoo Kim on 2/13/21.
//

#import "PlayScene.h"
#import "GameScene.h"
#import "EnemyNode.h"
#import "NSArray+Shuffled.h"

@interface PlayScene ()
@property SKSpriteNode * player;
@property NSArray<NSNumber *> *levelWaves;
@property NSUInteger levelPosition;
@property NSArray<NSValue *> *positions;
@property BOOL playerIsAlive;
@property SKLabelNode *scoreLabel;
@property (nonatomic) NSInteger score;
@property SKAudioNode *music;
@end

@implementation PlayScene

CGFloat const enemyOffsetX = 100;
static CGFloat const enemyStartX = 1100;

- (void)setup {
    self.player = [SKSpriteNode spriteNodeWithImageNamed:@"player"];
    self.levelWaves = @[
        [NSNumber numberWithUnsignedInteger:EnemyWaveSmallSmall],
        [NSNumber numberWithUnsignedInteger:EnemyWaveMediumSmall],
        [NSNumber numberWithUnsignedInteger:EnemyWaveLargeSmall],
        [NSNumber numberWithUnsignedInteger:EnemyWaveSmallMedium],
        [NSNumber numberWithUnsignedInteger:EnemyWaveMediumMedium],
        [NSNumber numberWithUnsignedInteger:EnemyWaveLargeMedium],
        [NSNumber numberWithUnsignedInteger:EnemyWaveSmallLarge],
        [NSNumber numberWithUnsignedInteger:EnemyWaveMediumLarge],
        [NSNumber numberWithUnsignedInteger:EnemyWaveLargeLarge],
        [NSNumber numberWithUnsignedInteger:EnemyWaveFlurrySmall],
        [NSNumber numberWithUnsignedInteger:EnemyWaveFlurryMedium],
        [NSNumber numberWithUnsignedInteger:EnemyWaveFlurryLarge]
    ];
    self.levelPosition = 0;
    self.positions = @[
        [NSValue valueWithCGPoint:CGPointMake(enemyStartX, 400)],
        [NSValue valueWithCGPoint:CGPointMake(enemyStartX, 300)],
        [NSValue valueWithCGPoint:CGPointMake(enemyStartX, 200)],
        [NSValue valueWithCGPoint:CGPointMake(enemyStartX, 100)],
        [NSValue valueWithCGPoint:CGPointMake(enemyStartX, 0)],
        [NSValue valueWithCGPoint:CGPointMake(enemyStartX, -100)],
        [NSValue valueWithCGPoint:CGPointMake(enemyStartX, -200)],
        [NSValue valueWithCGPoint:CGPointMake(enemyStartX, -300)],
        [NSValue valueWithCGPoint:CGPointMake(enemyStartX, -400)]
    ];
    self.playerIsAlive = YES;
    self.scoreLabel = [SKLabelNode labelNodeWithFontNamed:@"AvenirNext-Bold"];
    self.score = 0;
    
    // [SKAudioNode nodeWithFileNamed:@"cyborgNinja.mp3"]; 가 아님!!!
    self.music = [[SKAudioNode alloc] initWithFileNamed:@"cyborgNinja.mp3"];
}

- (void)setScore:(NSInteger)score {
    _score = score;
    self.scoreLabel.text = [NSString stringWithFormat:@"Score: %ld", score];
}

- (void)didMoveToView:(SKView *)view {
    [self setup];
    [super didMoveToView:view];
    self.anchorPoint = CGPointMake(0.5, 0.5);
    
    SKSpriteNode *background = [SKSpriteNode spriteNodeWithImageNamed:@"background"];
    background.zPosition = -2;
    background.blendMode = SKBlendModeReplace;
    [self addChild:background];
    
    SKEmitterNode *particles = [SKEmitterNode nodeWithFileNamed:@"starfield"];
    if (particles) {
        particles.position = CGPointMake(1080, 0);
        // 60초가 된 시점의 프레임부터 보여준다.
        [particles advanceSimulationTime:60];
        [self addChild:particles];
    }
    
    self.player.name = @"player";
    self.player.position = CGPointMake(-700, self.player.position.y);
    self.player.zPosition = 1;
    [self addChild:self.player];
    
    //
    
    // create a physics body for the player using its texture at its current size
    self.player.physicsBody = [SKPhysicsBody bodyWithTexture:self.player.texture size:[self.player.texture size]];
    
    // make this as being the player
    self.player.physicsBody.categoryBitMask = CollisionPlayerType;
    
    // tell it to bounce off enemies and enemy weapon
    self.player.physicsBody.collisionBitMask = CollisionEnemyType | CollisionEnemyWeaponType;
    
    // ask SpriteKit to tell us when it collides with enemies and their weapons
    self.player.physicsBody.contactTestBitMask = CollisionEnemyType | CollisionEnemyWeaponType;
    
    self.physicsWorld.gravity = CGVectorMake(0, 0);
    
    self.physicsWorld.contactDelegate = self;
    
    //
    
    self.scoreLabel.position = CGPointMake(0, 400);
    self.scoreLabel.zPosition = -1;
    self.scoreLabel.fontSize = 96;
    [self addChild:self.scoreLabel];
    self.score = 0;
    
    //
    
    [self addChild:self.music];
}

- (void)movePlayer:(CGPoint)delta {
    self.player.position = CGPointMake(self.player.position.x, self.player.position.y - (delta.y * 2));
    
    if (self.player.position.y < -500) {
        self.player.position = CGPointMake(self.player.position.x, -500);
    } else if (self.player.position.y > 500) {
        self.player.position = CGPointMake(self.player.position.x, 500);
    }
}

- (void)createWave {
    if (!self.playerIsAlive) return;
    if (self.levelPosition >= self.levelWaves.count) return;
    
    switch ([self.levelWaves[self.levelPosition] unsignedIntegerValue]) {
        case EnemyWaveSmallSmall: {
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[0] CGPointValue]
                                                   xOffset:enemyOffsetX * 2
                                              moveStraight:YES]];
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[4] CGPointValue]
                                                   xOffset:0
                                              moveStraight:YES]];
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[8] CGPointValue]
                                                   xOffset:enemyOffsetX * 2
                                              moveStraight:YES]];
            break;
        }
        case EnemyWaveMediumSmall: {
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[2] CGPointValue]
                                                   xOffset:0
                                              moveStraight:NO]];
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[1] CGPointValue]
                                                   xOffset:enemyOffsetX * 2
                                              moveStraight:NO]];
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[0] CGPointValue]
                                                   xOffset:enemyOffsetX * 4
                                              moveStraight:NO]];
            
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[6] CGPointValue]
                                                   xOffset:enemyOffsetX
                                              moveStraight:NO]];
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[7] CGPointValue]
                                                   xOffset:enemyOffsetX * 3
                                              moveStraight:NO]];
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[8] CGPointValue]
                                                   xOffset:enemyOffsetX * 5
                                              moveStraight:NO]];
            break;
        }
        case EnemyWaveLargeSmall: {
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[4] CGPointValue]
                                                   xOffset:0
                                              moveStraight:YES]];
            
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[3] CGPointValue]
                                                   xOffset:enemyOffsetX
                                              moveStraight:YES]];
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[2] CGPointValue]
                                                   xOffset:enemyOffsetX * 2
                                              moveStraight:YES]];
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[1] CGPointValue]
                                                   xOffset:enemyOffsetX * 3
                                              moveStraight:YES]];
            
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[5] CGPointValue]
                                                   xOffset:enemyOffsetX
                                              moveStraight:YES]];
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[6] CGPointValue]
                                                   xOffset:enemyOffsetX * 2
                                              moveStraight:YES]];
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[7] CGPointValue]
                                                   xOffset:enemyOffsetX * 3
                                              moveStraight:YES]];
            
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[2] CGPointValue]
                                                   xOffset:enemyOffsetX * 8
                                              moveStraight:NO]];
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[1] CGPointValue]
                                                   xOffset:enemyOffsetX * 8
                                              moveStraight:NO]];
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[0] CGPointValue]
                                                   xOffset:enemyOffsetX * 8
                                              moveStraight:NO]];
            
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[6] CGPointValue]
                                                   xOffset:enemyOffsetX * 9
                                              moveStraight:NO]];
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[7] CGPointValue]
                                                   xOffset:enemyOffsetX * 9
                                              moveStraight:NO]];
            [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                             startPosition:[self.positions[8] CGPointValue]
                                                   xOffset:enemyOffsetX * 9
                                              moveStraight:NO]];
            break;
        }
        default: {
            NSArray<NSValue *> *randomPositions = [self.positions shuffled];
            
            [randomPositions enumerateObjectsUsingBlock:^(NSValue * _Nonnull position, NSUInteger idx, BOOL * _Nonnull stop) {
                [self addChild:[[EnemyNode alloc] initWithType:@"enemy1"
                                                 startPosition:[position CGPointValue]
                                                       xOffset:enemyOffsetX * (CGFloat)(idx * 3)
                                                  moveStraight:YES]];
            }];
            
            break;
        }
    }
    
    self.levelPosition += 1;
}

- (void)update:(NSTimeInterval)currentTime {
    [super update:currentTime];
    
    for (SKNode *child in self.children) {
        if (CGRectGetMaxX(child.frame) < 0) {
            if (!CGRectIntersectsRect(self.frame, child.frame)) {
                [child removeFromParent];
            }
        }
    }
    
    NSPredicate *predicate = [NSPredicate predicateWithBlock:^BOOL(id  _Nullable evaluatedObject, NSDictionary<NSString *,id> * _Nullable bindings) {
        return [evaluatedObject isKindOfClass:[EnemyNode class]];
//        return ((EnemyNode *)evaluatedObject != nil);
    }];
    NSArray<EnemyNode *> *enemies = (NSArray<EnemyNode *> *)[self.children filteredArrayUsingPredicate:predicate];
    
    if (enemies.count == 0) [self createWave];
    
    //
    
    for (EnemyNode *enemy in enemies) {
        if ((enemy.lastFireTime + 1) < currentTime) {
            enemy.lastFireTime = currentTime;
            if (arc4random_uniform(2)) {
                [enemy fire];
            }
        }
    }
}

- (void)pressesBegan:(NSSet<UIPress *> *)presses withEvent:(UIPressesEvent *)event {
    if (!self.playerIsAlive) return;
    UIPress *press = [[presses allObjects] firstObject];
    if (press == nil) return;
    
    // check that the player pressed the play button
    if (press.type == UIPressTypePlayPause) {
        // create a weapon
        SKSpriteNode *shot = [SKSpriteNode spriteNodeWithImageNamed:@"playerWeapon"];
        shot.name = @"playerWeapon";
        
        // make it start from the player's position
        shot.position = self.player.position;
        
        // give it rectangular physics
        shot.physicsBody = [SKPhysicsBody bodyWithRectangleOfSize:shot.size];
        shot.physicsBody.categoryBitMask = CollisionPlayerWeaponType;
        shot.physicsBody.contactTestBitMask = CollisionEnemyType;
        shot.physicsBody.collisionBitMask = CollisionEnemyType;
        [self addChild:shot];
        
        // tell it to move to the other side of the screen then destory itself
        SKAction *movement = [SKAction moveTo:CGPointMake(1900, shot.position.y) duration:1];
        SKAction *sequence = [SKAction sequence:@[movement, [SKAction removeFromParent]]];
        
        // run it immediately
        [shot runAction:sequence];
        
        self.score -= 1;
        
        //
        
        [self runAction:[SKAction playSoundFileNamed:@"playerWeapon.wav" waitForCompletion:NO]];
    }
}

- (void)gameOver {
    self.playerIsAlive = NO;
    
    SKEmitterNode *explosion = [SKEmitterNode nodeWithFileNamed:@"explosion"];
    if (explosion) {
        explosion.position = self.player.position;
        [self addChild:explosion];
    }
    
    SKSpriteNode *gameOver = [SKSpriteNode spriteNodeWithImageNamed:@"gameOver"];
    [self addChild:gameOver];
    
    [self.music runAction:[SKAction stop]];
}

- (void)quitGame {
    GameScene *menu = [GameScene sceneWithSize:self.size];
    menu.anchorPoint = CGPointMake(0.5, 0.5);
    SKTransition *transition = [SKTransition doorsCloseVerticalWithDuration:1];
    [self.view presentScene:menu transition:transition];
    
    // tell the root view controller to update its focus
    [menu runAction:[SKAction waitForDuration:0.1] completion:^{
        [menu.view.window.rootViewController setNeedsFocusUpdate];
    }];
}

#pragma SKPhysicsContactDelegate

- (void)didBeginContact:(SKPhysicsContact *)contact {
    // make sure node colliding objects have SpriteKit nodes attached
    SKNode *nodeA = contact.bodyA.node;
    SKNode *nodeB = contact.bodyB.node;
    if ((nodeA == nil) || (nodeB == nil)) return;
    
    // if either node was the player, the game is over
    if ([nodeA.name isEqualToString:@"player"] || [nodeB.name isEqualToString:@"player"]) {
        if (!self.playerIsAlive) return;
        [self gameOver];
    } else {
        // create an explosion
        SKEmitterNode *explosion = [SKEmitterNode nodeWithFileNamed:@"explosion"];
        if (explosion) {
            // if the first object is the player weapon
            if ([nodeA.name isEqualToString:@"playerWeapon"]) {
                // position the explosion over the other object
                explosion.position = nodeB.position;
            } else {
                // otherwise position the object over the first object
                explosion.position = nodeA.position;
            }
            
            [self runAction:[SKAction playSoundFileNamed:@"explosion.wav" waitForCompletion:NO]];
            
            // add the explosion to the game
            [self addChild:explosion];
            
            self.score += 10;
        }
    }
    
    [nodeA removeFromParent];
    [nodeB removeFromParent];
}

@end

File: swift/Project8/Project8/PlayScene.swift
//
//  PlayScene.swift
//  Project8
//
//  Created by Jinwoo Kim on 2/12/21.
//

import SpriteKit

enum CollisionTypes: UInt32 {
    case player = 1
    case playerWeapon = 2
    case enemy = 4
    case enemyWeapon = 8
}

class PlayScene: SKScene {
    let player = SKSpriteNode(imageNamed: "player")
    
    let levelWaves: [EnemyWave] = [.smallSmall, .mediumSmall, .smallSmall, .largeSmall, .flurrySmall, .smallMedium, .mediumMedium, .largeMedium, .largeSmall, .mediumMedium, .smallLarge, .largeSmall, .largeLarge]
    var levelPosition = 0
    
    let enemyOffsetX: CGFloat = 100
    static let enemyStartX = 1100

    let positions = [
        CGPoint(x: enemyStartX, y: 400),
        CGPoint(x: enemyStartX, y: 300),
        CGPoint(x: enemyStartX, y: 200),
        CGPoint(x: enemyStartX, y: 100),
        CGPoint(x: enemyStartX, y: 0),
        CGPoint(x: enemyStartX, y: -100),
        CGPoint(x: enemyStartX, y: -200),
        CGPoint(x: enemyStartX, y: -300),
        CGPoint(x: enemyStartX, y: -400)
    ]
    
    var playerIsAlive = true
    
    var scoreLabel = SKLabelNode(fontNamed: "AvenirNext-Bold")
    var score = 0 {
        didSet {
            scoreLabel.text = "Score: \(score)"
        }
    }
    
    let music = SKAudioNode(fileNamed: "cyborgNinja.mp3")
    
    override func didMove(to view: SKView) {
        super.didMove(to: view)
        anchorPoint = CGPoint(x: 0.5, y: 0.5)
        
        let background = SKSpriteNode(imageNamed: "background")
        background.zPosition = -2
        background.blendMode = .replace
        addChild(background)
        
        if let particles = SKEmitterNode(fileNamed: "starfield") {
            particles.position = CGPoint(x: 1080, y: 0)
            // 60초가 된 시점의 프레임부터 보여준다.
            particles.advanceSimulationTime(60)
            addChild(particles)
        }
        
        player.name = "player"
        player.position.x = -700
        player.zPosition = 1
        addChild(player)
        
        //
        
        // create a physics body for the player using its texture at its current size
        player.physicsBody = SKPhysicsBody(texture: player.texture!, size: player.texture!.size())
        
        // make this as being the player
        player.physicsBody?.categoryBitMask = CollisionTypes.player.rawValue
        
        // tell it to bounce off enemies and enemy weapon
        player.physicsBody?.collisionBitMask = CollisionTypes.enemy.rawValue | CollisionTypes.enemyWeapon.rawValue
        
        // ask SpriteKit to tell us when it collides with enemies and their weapons
        player.physicsBody?.contactTestBitMask = CollisionTypes.enemy.rawValue | CollisionTypes.enemyWeapon.rawValue
        
        physicsWorld.gravity = .zero
        
        physicsWorld.contactDelegate = self
        
        //
        
        scoreLabel.position = CGPoint(x: 0, y: 400)
        scoreLabel.zPosition = -1
        scoreLabel.fontSize = 96
        addChild(scoreLabel)
        score = 0
        
        //
        
        addChild(music)
    }
    
    func movePlayer(_ delta: CGPoint) {
        player.position.y -= delta.y * 2
        
        if player.position.y < -500 {
            player.position.y = -500
        } else if player.position.y > 500 {
            player.position.y = 500
        }
    }
    
    func createWave() {
        guard playerIsAlive else { return }
        guard levelPosition < levelWaves.count else { return }

        switch levelWaves[levelPosition] {
        case .smallSmall:
            addChild(EnemyNode(type: "enemy1", startPosition: positions[0], xOffset: enemyOffsetX * 2, moveStraight: true))
            addChild(EnemyNode(type: "enemy1", startPosition: positions[4], xOffset: 0, moveStraight: true))
            addChild(EnemyNode(type: "enemy1", startPosition: positions[8], xOffset: enemyOffsetX * 2, moveStraight: true))

        case .mediumSmall:
            addChild(EnemyNode(type: "enemy1", startPosition: positions[2], xOffset: 0, moveStraight: false))
            addChild(EnemyNode(type: "enemy1", startPosition: positions[1], xOffset: enemyOffsetX * 2, moveStraight: false))
            addChild(EnemyNode(type: "enemy1", startPosition: positions[0], xOffset: enemyOffsetX * 4, moveStraight: false))

            addChild(EnemyNode(type: "enemy1", startPosition: positions[6], xOffset: enemyOffsetX, moveStraight: false))
            addChild(EnemyNode(type: "enemy1", startPosition: positions[7], xOffset: enemyOffsetX * 3, moveStraight: false))
            addChild(EnemyNode(type: "enemy1", startPosition: positions[8], xOffset: enemyOffsetX * 5, moveStraight: false))

        case .largeSmall:
            addChild(EnemyNode(type: "enemy1", startPosition: positions[4], xOffset: 0, moveStraight: true))

            addChild(EnemyNode(type: "enemy1", startPosition: positions[3], xOffset: enemyOffsetX, moveStraight: true))
            addChild(EnemyNode(type: "enemy1", startPosition: positions[2], xOffset: enemyOffsetX * 2, moveStraight: true))
            addChild(EnemyNode(type: "enemy1", startPosition: positions[1], xOffset: enemyOffsetX * 3, moveStraight: true))

            addChild(EnemyNode(type: "enemy1", startPosition: positions[5], xOffset: enemyOffsetX, moveStraight: true))
            addChild(EnemyNode(type: "enemy1", startPosition: positions[6], xOffset: enemyOffsetX * 2, moveStraight: true))
            addChild(EnemyNode(type: "enemy1", startPosition: positions[7], xOffset: enemyOffsetX * 3, moveStraight: true))

            addChild(EnemyNode(type: "enemy1", startPosition: positions[2], xOffset: enemyOffsetX * 8, moveStraight: false))
            addChild(EnemyNode(type: "enemy1", startPosition: positions[1], xOffset: enemyOffsetX * 8, moveStraight: false))
            addChild(EnemyNode(type: "enemy1", startPosition: positions[0], xOffset: enemyOffsetX * 8, moveStraight: false))

            addChild(EnemyNode(type: "enemy1", startPosition: positions[6], xOffset: enemyOffsetX * 9, moveStraight: false))
            addChild(EnemyNode(type: "enemy1", startPosition: positions[7], xOffset: enemyOffsetX * 9, moveStraight: false))
            addChild(EnemyNode(type: "enemy1", startPosition: positions[8], xOffset: enemyOffsetX * 9, moveStraight: false))

        default:
            let randomPositions = positions.shuffled()

            for (index, position) in randomPositions.enumerated() {
                addChild(EnemyNode(type: "enemy1", startPosition: position, xOffset: enemyOffsetX * CGFloat(index * 3), moveStraight: true))
            }
        }
        
        levelPosition += 1
    }
    
    override func update(_ currentTime: TimeInterval) {
        super.update(currentTime)
        
        for child in children {
            if child.frame.maxX < 0 {
                if !frame.intersects(child.frame) {
                    child.removeFromParent()
                }
            }
        }
        
        let enemies = children.compactMap { $0 as? EnemyNode }
        
        if enemies.isEmpty {
            createWave()
        }
        
        //
        
        for enemy in enemies {
            if enemy.lastFireTime + 1 < currentTime {
                enemy.lastFireTime = currentTime
                if Bool.random() {
                    enemy.fire()
                }
            }
        }
    }
    
    override func pressesBegan(_ presses: Set<UIPress>, with event: UIPressesEvent?) {
        guard playerIsAlive else { return }
        guard let press = presses.first else { return }
        
        // check that the player pressed the play button
        if press.type == .playPause {
            // create a weapon
            let shot = SKSpriteNode(imageNamed: "playerWeapon")
            shot.name = "playerWeapon"
            
            // make it start from the player's position
            shot.position = player.position
            
            // give it rectangular physics
            shot.physicsBody = SKPhysicsBody(rectangleOf: shot.size)
            shot.physicsBody?.categoryBitMask = CollisionTypes.playerWeapon.rawValue
            shot.physicsBody?.contactTestBitMask = CollisionTypes.enemy.rawValue
            shot.physicsBody?.collisionBitMask = CollisionTypes.enemy.rawValue
            addChild(shot)
            
            // tell it to move to the other side of the screen then destroy itself
            let movement = SKAction.move(to: CGPoint(x: 1900, y: shot.position.y), duration: 1)
            let sequence = SKAction.sequence([movement, SKAction.removeFromParent()])
            
            // run it immediately
            shot.run(sequence)
            
            score -= 1
            
            //
            
            run(SKAction.playSoundFileNamed("playerWeapon.wav", waitForCompletion: false))
        }
    }
    
    func gameOver() {
        playerIsAlive = false
        
        if let explosion = SKEmitterNode(fileNamed: "explosion") {
            explosion.position = player.position
            addChild(explosion)
        }
        
        let gameOver = SKSpriteNode(imageNamed: "gameOver")
        addChild(gameOver)
        
        music.run(SKAction.stop())
    }
    
    func quitGame() {
        let menu = GameScene(size: size)
        menu.anchorPoint = CGPoint(x: 0.5, y: 0.5)
        let transition = SKTransition.doorsCloseVertical(withDuration: 1)
        view?.presentScene(menu, transition: transition)
        
        // tell the root view controller to update its focus
        menu.run(SKAction.wait(forDuration: 0.1)) {
            menu.view?.window?.rootViewController?.setNeedsFocusUpdate()
        }
    }
}

extension PlayScene: SKPhysicsContactDelegate {
    func didBegin(_ contact: SKPhysicsContact) {
        // make sure node colliding objects have SpriteKit nodes attached
        guard let nodeA = contact.bodyA.node else { return }
        guard let nodeB = contact.bodyB.node else { return }
        
        // if either node was the player, the game is over
        if nodeA.name == "player" || nodeB.name == "player" {
            guard playerIsAlive else { return }
            gameOver()
        } else {
            // create an explosion
            if let explosion = SKEmitterNode(fileNamed: "explosion") {
                // if the first object is the player weapon
                if nodeA.name == "playerWeapon" {
                    // position the explosion over the other object
                    explosion.position = nodeB.position
                } else {
                    // otherwise position the object over the first object
                    explosion.position = nodeA.position
                }
                
                run(SKAction.playSoundFileNamed("explosion.wav", waitForCompletion: false))
                
                // add the explosion to the game
                addChild(explosion)
                
                score += 10
            }
        }
        
        nodeA.removeFromParent()
        nodeB.removeFromParent()
    }
}
